File: java/treap/Naive.java
package treap;

import java.util.*;
import java.lang.*;

public final class Naive {
    static final class SplitResult {
        SplitResult(Node lower, Node equal, Node greater) {
            this.lower = lower;
            this.equal = equal;
            this.greater = greater;
        }

        public Node lower;
        public Node equal;
        public Node greater;
    }

    static final class NodePair {
        NodePair(Node first, Node second) {
            this.first = first;
            this.second = second;
        }

        public Node first;
        public Node second;
    }

    static final class Node {
        public static Random random = new Random();

        Node(int x) {
            this.x = x;
        }

        int x;
        int y = random.nextInt();
        Node left = null;
        Node right = null;

        public static Node merge(Node lower, Node greater) {
            if (lower == null)
                return greater;

            if (greater == null)
                return lower;

            if (lower.y < greater.y) {
                lower.right = merge(lower.right, greater);
                return lower;
            } else {
                greater.left = merge(lower, greater.left);
                return greater;
            }
        }

        public static NodePair splitBinary(Node orig, int value) {
            if (orig == null)
                return new NodePair(null, null);

            if (orig.x < value) {
                NodePair splitPair = splitBinary(orig.right, value);
                orig.right = splitPair.first;
                return new NodePair(orig, splitPair.second);
            } else {
                NodePair splitPair = splitBinary(orig.left, value);
                orig.left = splitPair.second;
                return new NodePair(splitPair.first, orig);
            }
        }

        public static Node merge(Node lower, Node equal, Node greater) {
            return merge(merge(lower, equal), greater);
        }

        public static SplitResult split(Node orig, int value) {
            NodePair lowerOther = splitBinary(orig, value);
            NodePair equalGreater = splitBinary(lowerOther.second, value + 1);
            return new SplitResult(lowerOther.first, equalGreater.first, equalGreater.second);
        }
    }

    static final class Tree {
        public boolean hasValue(int x) {
            SplitResult splited = Node.split(mRoot, x);
            boolean res = splited.equal != null;
            mRoot = Node.merge(splited.lower, splited.equal, splited.greater);
            return res;
        }

        public void insert(int x) {
            SplitResult splited = Node.split(mRoot, x);
            if (splited.equal == null)
                splited.equal = new Node(x);
            mRoot = Node.merge(splited.lower, splited.equal, splited.greater);
        }

        public void erase(int x) {
            SplitResult splited = Node.split(mRoot, x);
            mRoot = Node.merge(splited.lower, splited.greater);
        }

        private Node mRoot = null;
    }

    public static void main(String[] args) {
        long start = System.currentTimeMillis();

        Tree tree = new Tree();
        int cur = 5;
        int res = 0;

        for (int i = 1; i < 1000000; i++) {
            int a = i % 3;
            cur = (cur * 57 + 43) % 10007;
            if (a == 0) {
                tree.insert(cur);
            } else if (a == 1) {
                tree.erase(cur);
            } else if (a == 2) {
                boolean hasVal = tree.hasValue(cur);
                if (hasVal)
                    res++;
            }
        }
        System.out.println(res);

        long end = System.currentTimeMillis();
        long duration = end - start;

        System.out.println("Execution time: " + duration + "ms");
    }
}

File: lua/treap/Naive.lua
math.randomseed(os.time())

Node = {}
Node.__index = Node

function Node:new(x)
    local instance = setmetatable({}, self)
    instance.x = x
    instance.y = math.random()
    instance.left = nil
    instance.right = nil
    return instance
end

function merge(lower, greater)
    if lower == nil then
        return greater
    end

    if greater == nil then
        return lower
    end

    if lower.y < greater.y then
        lower.right = merge(lower.right, greater)
        return lower
    else
        greater.left = merge(lower, greater.left)
        return greater
    end
end

function splitBinary(orig, value)
    if orig == nil then
        return nil, nil
    end

    if orig.x < value then
        local rightLower, rightGreater = splitBinary(orig.right, value)
        orig.right = rightLower
        return orig, rightGreater
    else
        local leftLower, leftGreater = splitBinary(orig.left, value)
        orig.left = leftGreater
        return leftLower, orig
    end
end

function merge3(lower, equal, greater)
    return merge(merge(lower, equal), greater)
end

SplitResult = {}
SplitResult.__index = SplitResult

function SplitResult:new(lower, equal, greater)
    local instance = setmetatable({}, self)
    instance.lower = lower
    instance.equal = equal
    instance.greater = greater
    return instance
end

function split(orig, value)
    local lower, equalGreater = splitBinary(orig, value)
    local equal, greater = splitBinary(equalGreater, value + 1)
    return SplitResult:new(lower, equal, greater)
end

Tree = {}
Tree.__index = Tree

function Tree:new()
    local instance = setmetatable({}, self)
    instance.root = nil
    return instance
end

function Tree:hasValue(x)
    local splited = split(self.root, x)
    local res = splited.equal ~= nil
    self.root = merge3(splited.lower, splited.equal, splited.greater)
    return res
end

function Tree:insert(x)
    local splited = split(self.root, x)
    if splited.equal == nil then
        splited.equal = Node:new(x)
    end
    self.root = merge3(splited.lower, splited.equal, splited.greater)
end

function Tree:erase(x)
    local splited = split(self.root, x)
    self.root = merge(splited.lower, splited.greater)
end

function main()
    local tree = Tree:new()
    local cur = 5
    local res = 0

    for i = 1, 999999 do
        local a = i % 3
        cur = (cur * 57 + 43) % 10007

        if a == 0 then
            tree:insert(cur)
        elseif a == 1 then
            tree:erase(cur)
        elseif a == 2 then
            res = res + (tree:hasValue(cur) and 1 or 0)
        end
    end

    print(res)
end

local startTimeMs = math.floor(os.clock() * 1000)
main()
local endTimeMs = math.floor(os.clock() * 1000)
local durationMs = endTimeMs - startTimeMs

print("Execution time: " .. durationMs .. "ms")
